### DISCUSSION TITLE
Adding an alias to zsh-abbr via fzf

### DISCUSSION THREAD

**Discussion by LangLangBart** [olets/zsh-abbr#94] [Q&A] [answered]
I have a question when adding a selected alias via `fzf`

```sh
# select any existing aliases from a list with ‘fzf’ and add them to 'zsh-abbr'
alias | fzf --multi | while IFS= read -r line; do abbr add --force "$line"; done
```

Once the alias is added, the extended alias starts and ends with single quotes. See GIF below for more clarity.

<img src="https://github.com/olets/zsh-abbr/assets/92653266/1e97f753-aa21-4696-bf05-4cf1d2fa01ff" width="600">

if you add `set -x` the problem becomes obvious
Short notation | Long notation | Result
-- | -- | --
set -v | set -o verbose | Prints shell input lines as they are read.
set -x | set -o xtrace | Print command traces before executing command.

```sh
alias | fzf --multi | while IFS= read -r line; do set -x; abbr add --force "$line";set +x; done
# +zsh:18> abbr add --force 'agif='\''agg --font-dir $HOME/Library/Fonts --font-family "MesloLGS NF" --font-size 18 --fps-cap 20 --theme asciinema --speed 1.5 --renderer resvg $HOME/Desktop/asciinema.cast $HOME/Desktop/them.gif'\'
# +abbr:1> emulate -LR zsh
# Redefined the regular user abbreviation `agif`
# +zsh:18> set +x
```

**Question**: Is this an issue with `fzf` or `zsh-abbr` ?

The following works as a workaround, by removing the single quotes.
```sh
alias | fzf --multi | while IFS= read -r line; do abbr add --force "$(sed <<<"$line" 's/=./=/;s/.$//')"; done
```

---

PS: The tool you developed is very useful and has been incorporated into my workflow.

**Comment by LangLangBart** [comment, marked as answer]
- `eval` was needed
  - ref: [Shell-Builtin-Commands.html#index-eval](https://zsh.sourceforge.io/Doc/Release/Shell-Builtin-Commands.html#index-eval) 
  - ref: [A User's Guide to the Z-Shell by Peter Stephenson - 5.1: Quoting](https://zsh.sourceforge.io/Guide/zshguide05.html#l112)

```sh
alias | fzf --multi | while IFS= read -r line; do abbr add --force "$(eval print -r -- "$line")"; done
```

**Comment by olets** [comment]
Is the result different from the result of zsh-abbr's import-aliases? https://zsh-abbr.olets.dev/commands.html#import-aliases

**Reply by LangLangBart** [reply]
> _Is the result different from the result of zsh-abbr's import-aliases? https://zsh-abbr.olets.dev/commands.html#import-aliases_

My `fzf` function seems to work mostly fine, noticed some trouble with quotes in the form of `$’...’` (aka [ANSI-C Quoting](https://www.gnu.org/software/bash/manual/html_node/ANSI_002dC-Quoting.html), [C-style quotes](https://zsh.sourceforge.io/Doc/Release/Expansion.html#Overview) or [POSIX quotes](https://zsh.sourceforge.io/Guide/zshguide05.html#l115)). It seems that `fzf` automatically inserts them at the beginning of the line if it decides they are necessary. For individually selected items in the `fzf` list it does work, but part of a multi selection it doesn't. Such aliases are skipped in my function.

```sh
# example of an alias with $'...'
alias color='for i in {0..255}; do print -Pn "%K{$i}  %k%F{$i}${(l:3::0:)i}%f " ${${(M)$((i%6)):#3}:+'$'\n}; done'
```

<img src="https://github.com/olets/zsh-abbr/assets/92653266/255556f8-3f21-47c8-8935-815eea8d7d80" width="800">

```sh
# debug message for aliases with $'...' prefixes (olet is the name of my function)
+olet:11> eval print -r 'color=$'\''for i in {0..255}; do print -Pn "%K{$i}  %k%F{$i}${(l:3::0:)i}%f " ${${(M)$((i%6)):#3}:+\n}; done'\'
+(eval):1> print -r $'color=for i in {0..255}; do print -Pn "%K{$i}  %k%F{$i}${(l:3::0:)i}%f " ${${(M)$((i%6)):#3}:+\n}; done'
+olet:11> abbr add --force $'color=for i in {0..255}; do print -Pn "%K{$i}  %k%F{$i}${(l:3::0:)i}%f " ${${(M)$((i%6)):#3}:+\n}; done'
+abbr:1> emulate -LR zsh
Redefined the regular user abbreviation `color`

# 'color' is only partially expanded
"for i in {0..255}; do print -Pn \"%K{\$i}  %k%F{\$i}\${(l:3::0:)i}%f \" \${\${(M)\$((i%6)):#3}:+ 
```
---

Your setup with `abbr import-aliases` seems to work more reliably for such edge cases, as seen above.

---

PS: I knew the `abbr import-aliases` utility because of the good documentation. I just didn't want to add all aliases, because I got used to some never being expanded and other aliases being very long, and it looks bad on the command line when you see their long expansion. I read about the <kbd>⌃ Control</kbd> + <kbd>Space</kbd> hotkey, but pressing an additional hotkey for certain aliases is more cumbersome and this particular hotkey is already assigned to toggle the [egoist/devdocs-desktop](https://github.com/egoist/devdocs-desktop/) app.
